File: 2021/day02/php/solution.php
<?php

function parseInput(): array {
    $inputPath = __DIR__ . '/../input.txt';
    $lines = file($inputPath, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    $commands = [];
    foreach ($lines as $line) {
        $parts = explode(' ', trim($line));
        $commands[] = [$parts[0], (int)$parts[1]];
    }
    return $commands;
}

function part1(array $commands): int {
    $horizontal = 0;
    $depth = 0;
    foreach ($commands as [$cmd, $val]) {
        switch ($cmd) {
            case 'forward':
                $horizontal += $val;
                break;
            case 'down':
                $depth += $val;
                break;
            case 'up':
                $depth -= $val;
                break;
        }
    }
    return $horizontal * $depth;
}

function part2(array $commands): int {
    $horizontal = 0;
    $depth = 0;
    $aim = 0;
    foreach ($commands as [$cmd, $val]) {
        switch ($cmd) {
            case 'forward':
                $horizontal += $val;
                $depth += $aim * $val;
                break;
            case 'down':
                $aim += $val;
                break;
            case 'up':
                $aim -= $val;
                break;
        }
    }
    return $horizontal * $depth;
}

$commands = parseInput();
echo "Part 1: " . part1($commands) . "\n";
echo "Part 2: " . part2($commands) . "\n";

File: 2021/day02/rust/solution.rs
use std::fs;
use std::path::Path;

fn parse_input() -> Vec<(String, i64)> {
    let input_path = Path::new(file!()).parent().unwrap().join("../input.txt");
    let content = fs::read_to_string(input_path).expect("Failed to read input file");

    content
        .lines()
        .filter(|line| !line.trim().is_empty())
        .map(|line| {
            let parts: Vec<&str> = line.split_whitespace().collect();
            (parts[0].to_string(), parts[1].parse().unwrap())
        })
        .collect()
}

fn part1(commands: &[(String, i64)]) -> i64 {
    let mut horizontal = 0;
    let mut depth = 0;

    for (cmd, val) in commands {
        match cmd.as_str() {
            "forward" => horizontal += val,
            "down" => depth += val,
            "up" => depth -= val,
            _ => {}
        }
    }

    horizontal * depth
}

fn part2(commands: &[(String, i64)]) -> i64 {
    let mut horizontal = 0;
    let mut depth = 0;
    let mut aim = 0;

    for (cmd, val) in commands {
        match cmd.as_str() {
            "forward" => {
                horizontal += val;
                depth += aim * val;
            }
            "down" => aim += val,
            "up" => aim -= val,
            _ => {}
        }
    }

    horizontal * depth
}

fn main() {
    let commands = parse_input();
    println!("Part 1: {}", part1(&commands));
    println!("Part 2: {}", part2(&commands));
}
